src/main/java/com/atlassian/db/replica/internal/ReplicaStatement.java

```diff
@@ -600,6 +663,9 @@ private boolean isFunctionCall(String sql) {
             return false;
         }
         final String potentialMethodName = mayContainFunction.split(methodBracketStart)[0];
+        if (standardReadOnlyFunctions.contains(potentialMethodName) || readOnlyFunctions.contains(potentialMethodName)) {
```

Comment:
`potentialMethodName` sounds fuzzy, but the `readOnlyFunctions` is a strict `contains`. Does this actually catch function names in a more messy query?

- [ ] test a case where first function is RO, but second is RW

Reply:
I know it's technically possible to have cases like this, but please find one from the production code. I don't think safety for the cases that may never occur is worth the extra complexity.

>  test with lowercase and uppercase function names

Lower and upper case functions are supported for the built-in functions, and it's covered. Supporting it for user-defined functions is up to the library user. If someone uses both upper and lower case invocations of the same method in the codebase, then both ways need to be registered.

Reply:
👍 case sensitivity is indeed covered:
```java
        connection.prepareStatement("SELECT max(user.id) FROM user").executeQuery();
        connection.prepareStatement("SELECT MAX(user.id) FROM user").executeQuery();
```

Reply:
> please find one from the production code

I cannot find it. Perhaps sanitizing it and putting it in the test would not be such a bad idea.

> I don't think safety for the cases that may never occur is worth the extra complexity.

How do we know these cases do not occur right now? Is a mix of RO and RW functions unrealistic? If so, why?

Reply:
> How do we know these cases do not occur right now? Is a mix of RO and RW functions unrealistic? If so, why?

We will know because we'll see an exception if we try to run RW operation on the replica. If it's realistic, please find one case like this in the logs we have.

Reply:
We see lots of exceptions though, tens, hundreds, thousands, depending on the day.

Reply:
But ok, one step at a time.
If these exceptions are due to insufficient function detection, then this PR does not make that worse.